File: leetcode/leetcode_problems/LinkedList.py
'''
LinkedList problems and solutions
'''

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class LinkedList:
    def __init__(self, data=None):
        self.head = None
        if data and len(data) > 0:
            self.head = ListNode(data[0])
            current = self.head
            for i in range(1, len(data)):
                current.next = ListNode(data[i])
                current = current.next
        
    def append(self, val):
        new_node = ListNode(val)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = new_node
        
    def insert(self, index, val):
        if index == 0:
            new_node = ListNode(val)
            new_node.next = self.head
            self.head = new_node
            return
        new_node = ListNode(val)
        current = self.head
        for _ in range(index - 1):
            if current is None:
                raise IndexError("Index out of range")
            current = current.next
        if current is None:
            raise IndexError("Index out of range")
        new_node.next = current.next
        current.next = new_node
        
    def delete(self, index):
        if self.head is None:
            raise IndexError("Index out of range")
        if index == 0:
            self.head = self.head.next
            return
        current = self.head
        for _ in range(index - 1):
            if current is None or current.next is None:
                raise IndexError("Index out of range")
            current = current.next
        if current.next is None:
            raise IndexError("Index out of range")
        current.next = current.next.next
        
    def print(self):
        current = self.head
        while current is not None:
            print(current.val, end=" ")
            current = current.next
        print()
        
    def reverse(self):
        prev = None
        current = self.head
        while current is not None:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        self.head = prev
        
    def find(self, val):
        current = self.head
        while current is not None:
            if current.val == val:
                return current
            current = current.next
        return None
    
    def length(self):
        current = self.head
        length = 0
        while current is not None:
            length += 1
            current = current.next
        return length
    
    def is_empty(self):
        return self.head is None
        
    def clear(self):
        self.head = None
        
    def get(self, index):
        current = self.head
        for _ in range(index):
            if current is None:
                raise IndexError("Index out of range")
            current = current.next
        if current is None:
            raise IndexError("Index out of range")
        return current.val
        
    def set(self, index, val):
        current = self.head
        for _ in range(index):
            if current is None:
                raise IndexError("Index out of range")
            current = current.next
        if current is None:
            raise IndexError("Index out of range")
        current.val = val

if __name__ == "__main__":
    linked_list = LinkedList([1, 2, 3, 4, 5])
    linked_list.print()
    print(linked_list.get(2))
    linked_list.set(2, 10)
    linked_list.print()
    print(linked_list.find(10))
    print(linked_list.length())
    print(linked_list.is_empty())
    linked_list.clear()
    linked_list.print()

File: leetcode/leetcode_problems/LinkedList.java
/*
LinkedList problems and solutions
 */

class LinkedList {
    // Define ListNode class
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    // Append ListNode method
    public void appendNode(ListNode head, int val) {
        ListNode newNode = new ListNode(val);
        ListNode current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    // Insert new ListNode after value method
    public void insertAfter(ListNode head, int targetVal, int newVal) {
        ListNode current = head;
        while (current != null) {
            if (current.val == targetVal) {
                ListNode newNode = new ListNode(newVal);
                newNode.next = current.next;
                current.next = newNode;
                return;
            }
            current = current.next;
        }
    }
    
    // Delete ListNode method
    public ListNode deleteNode(ListNode head, int val) {
        if (head == null) return null;
        if (head.val == val) return head.next;
        ListNode current = head;
        while (current.next != null) {
            if (current.next.val == val) {
                current.next = current.next.next;
                return head;
            }
            current = current.next;
        }
        return head;
    }

    // Print ListNode method
    public void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    // Example method: Reverse a linked list
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }
        return prev;
    }
}